/**
 * Copyright 2012 Carnegie Institution for Science. All rights reserved.
 */
// Generated by AndroMDA Poesys/DB cartridge: DO NOT EDIT DIRECTLY.
// Template: AbstractDbDto.vsl

package org.tair.db.locusdetail;


import com.poesys.db.pk.IPrimaryKey;

import com.poesys.db.dto.AbstractDto;
import com.poesys.db.dto.DtoStatusException;


/**
 * <p>
 * A data-access layer data-transfer object for the LocusGeneModelCdna. This class
 * is an abstract class that contains AndroMDA generated code; change nothing
 * in this class. Instead, override any methods in the concrete subclass
 * generated in the same package. AndroMDA will overwrite this class each time
 * you run it but will never overwrite the concrete subclass.
 * </p>
 * <p>
 * A complementary DNA (cDNA) sequence associated with the locus gene model. cDNA
 * is single-stranded DNA made in the laboratory from a messenger RNA template.
 * Biologists use this form of DNA as a probe in the structural mapping of a
 * chromosome.
 * </p>
 * <p>
 * Stereotypes:
 * </p>
 * <ul>
 *     <li>CompositeKey</li>
 *     <li>Persistent</li>
 * </ul>
 * <p>
 * Class is read/write: false
 * </p>
 *
 * @author Poesys/DB Cartridge
 */
public abstract class AbstractLocusGeneModelCdna extends AbstractDto implements ILocusGeneModelCdna {
  /** Default serial version UID for the Serializable DTO */
  private static final long serialVersionUID = 1L;
  // Setter strategy nested classes for single-object associations
  
  /**
   * Nested class that manages the geneModel association data
   *
   * Source: AddToOneAssociationRequiredObjectProperties
   *
   * @author Poesys/DB Cartridge
   */
  private class QueryGeneModelSetter extends com.poesys.db.dto.AbstractObjectSetter<org.tair.db.locusdetail.ILocusGeneModel> {
    private static final long serialVersionUID = 1L;
    
    /**
     * Create a QueryGeneModelSetter object.
     */
    public QueryGeneModelSetter() {
      super("org.tair.db.locusdetail", 2147483647);
    }

    @Override
    protected String getClassName() {
      return org.tair.db.locusdetail.LocusGeneModel.class.getName();
    }

    @Override
    protected IPrimaryKey getKey() {
      return geneModelKey;
    }

    @Override
    protected com.poesys.db.dao.query.IKeyQuerySql<org.tair.db.locusdetail.ILocusGeneModel> getSql() {
      return new org.tair.db.locusdetail.sql.QueryLocusGeneModel();
    }

    @Override
    protected void set(org.tair.db.locusdetail.ILocusGeneModel dto) {
      // No status change, this is just filling in the object data.
      geneModel = dto;
    }

    @Override
    public boolean isSet() {
      // Object is set if the associated geneModel is not null
      return geneModel != null;
    }
  }


  /**
   * Foreign key object used by QueryGeneModelSetter nested class to query object
   */
  private IPrimaryKey geneModelKey;
  
  /**
   * Set the foreign key geneModelKey. This has package access to enable
   * the subsystem factory getData method to call this method to set the key
   * by creating it from the queried result set.
   *
   * @param geneModelKey the foreign key
   */
  void setGeneModelKey(IPrimaryKey geneModelKey) {
    this.geneModelKey = geneModelKey;
  }

  // Setter strategy nested classes for multiple-object associations

  /**
   * Query setter for querying nested aliases
   *
   * Source: AddToManyChildCollectionProperties
   *
   * @see org.tair.db.locusdetail.sql.QueryAliasesByLocusGeneModelCdna
   */
  private class QueryAliasesSetter 
      extends com.poesys.db.dto.AbstractListSetter<org.tair.db.locusdetail.ILocusGeneModelCdnaAlias, ILocusGeneModelCdna, java.util.Collection<org.tair.db.locusdetail.ILocusGeneModelCdnaAlias>> {
    private static final long serialVersionUID = 1L;
    private static final int FETCH_SIZE = 10;

    /**
     * Create a QueryAliasesSetter object.
     */
    public QueryAliasesSetter() {
      super("org.tair.db.locusdetail", 2147483647);
    }

    @Override
    protected String getClassName() {
      return org.tair.db.locusdetail.LocusGeneModelCdnaAlias.class.getName();
    }

    @Override
    protected int getFetchSize() {
      return FETCH_SIZE;
    }

    @Override
    protected ILocusGeneModelCdna getParametersDto() {
      return AbstractLocusGeneModelCdna.this;
    }

    @Override
    protected com.poesys.db.dao.query.IParameterizedQuerySql<org.tair.db.locusdetail.ILocusGeneModelCdnaAlias, ILocusGeneModelCdna> getSql() {
      return new org.tair.db.locusdetail.sql.QueryAliasesByLocusGeneModelCdna();
    }

    @Override
    protected void set(java.util.Collection<org.tair.db.locusdetail.ILocusGeneModelCdnaAlias> list) {
      // No status change; this is just filling in the object data.
      aliases = list;
      // Add the primary keys to the serialized key list if there are any.
      if (aliases != null) {
        for (com.poesys.db.dto.IDbDto object : aliases) {
          aliasesKeys.add(object.getPrimaryKey());
        }
      }
    }

    @Override
    public boolean isSet() {
      // Object is set if the associated aliases list is not null
      return aliases != null;
    }
  }

  /**
   * Read-Object setter for de-serializing nested aliases collection
   *
   * Source: AddToManyChildCollectionProperties
   *
   * @see org.tair.db.locusdetail.sql.QueryLocusGeneModelCdnaAlias
   */
  private class ReadAliasesSetter 
      extends com.poesys.db.dto.AbstractCollectionReadSetter<org.tair.db.locusdetail.ILocusGeneModelCdnaAlias> {
    private static final long serialVersionUID = 1L;

    /**
     * Create a ReadAliasesSetter object to read the aliases collection.
     */
    public ReadAliasesSetter() {
      super("org.tair.db.locusdetail", 2147483647);
    }

    @Override
    protected String getClassName() {
      return org.tair.db.locusdetail.LocusGeneModelCdnaAlias.class.getName();
    }

    @Override
    protected java.util.Collection<org.tair.db.locusdetail.ILocusGeneModelCdnaAlias> getObjectCollection() {
      return aliases;
    }

    @Override
    protected java.util.List<IPrimaryKey> getPrimaryKeys() {
      return aliasesKeys;
    }

    @Override
    protected com.poesys.db.dao.query.IKeyQuerySql<org.tair.db.locusdetail.ILocusGeneModelCdnaAlias> getSql() {
      return new org.tair.db.locusdetail.sql.QueryLocusGeneModelCdnaAlias();
    }

    @Override
    protected void set(java.util.Collection<org.tair.db.locusdetail.ILocusGeneModelCdnaAlias> collection) {
      aliases = collection;
    }
  }
   
  /**
   * Create an empty LocusGeneModelCdna for use in building a new object. The 
   * concrete subclass must call this constructor.
   */
  public AbstractLocusGeneModelCdna() {
    // Setter arrays
    querySetters = new java.util.concurrent.CopyOnWriteArrayList<com.poesys.db.dto.ISet>();
    insertSetters = new java.util.concurrent.CopyOnWriteArrayList<com.poesys.db.dto.ISet>();
    preSetters = new java.util.concurrent.CopyOnWriteArrayList<com.poesys.db.dto.ISet>();
    postSetters = new java.util.concurrent.CopyOnWriteArrayList<com.poesys.db.dto.ISet>();
    readObjectSetters = new java.util.concurrent.CopyOnWriteArrayList<com.poesys.db.dto.ISet>();

    // Add the setters for the geneModel property.
    querySetters.add(new QueryGeneModelSetter());
    
    // Add the many-to-many collection setters for the aliases property.
    querySetters.add(new QueryAliasesSetter());
    readObjectSetters.add(new ReadAliasesSetter());
  }

  /**
   * Create a LocusGeneModelCdna. The concrete subclass must call this constructor.
   *
   * @param key the primary key of the LocusGeneModelCdna
   * @param geneModelName composite super-key attribute that uniquely identifies child combined with child sub-key and any other parent super-keys
   * @param locusTairObjectId composite super-key attribute that uniquely identifies child combined with child sub-key and any other parent super-keys
   * @param sequenceId the unique identifier for the cDNA
   * @param sequenceName the sequence name
   * @param icAccession the unique identifier at any of the major international genomics data centers
for the cDNA sequence
   * @param vectorType the kind of vector associated with the cDNA:
BAC
BiBAC
P1
TAC
YAC
cosmid
lambda
plasmid
unknown
   * @param fullLengthCds whether this cDNA is a "full-length" cDNA that contains the entire coding region
of the gene
   * @param length the length of the cDNA sequence, unit is bp (base pairs)
   * @param start the starting position on the chromosome in unit bp (base pairs)
   * @param end the ending position on the chromosome in unit bp (base pairs)
   */
  public AbstractLocusGeneModelCdna(IPrimaryKey key, java.lang.String geneModelName, java.lang.Long locusTairObjectId, java.lang.Long sequenceId, java.lang.String sequenceName, java.lang.String icAccession, java.lang.String vectorType, java.lang.Boolean fullLengthCds, java.lang.Long length, java.lang.Long start, java.lang.Long end) {
    this.key = key;

    this.geneModelName = geneModelName;

    if (geneModelName == null) {
      throw new com.poesys.db.InvalidParametersException("geneModelName is required for " + key.getValueList());
    }
    
    this.locusTairObjectId = locusTairObjectId;

    if (locusTairObjectId == null) {
      throw new com.poesys.db.InvalidParametersException("locusTairObjectId is required for " + key.getValueList());
    }
    
    this.sequenceId = sequenceId;

    if (sequenceId == null) {
      throw new com.poesys.db.InvalidParametersException("sequenceId is required for " + key.getValueList());
    }
    
    this.sequenceName = sequenceName;

    if (sequenceName == null) {
      throw new com.poesys.db.InvalidParametersException("sequenceName is required for " + key.getValueList());
    }
    
    this.icAccession = icAccession;

    if (icAccession == null) {
      throw new com.poesys.db.InvalidParametersException("icAccession is required for " + key.getValueList());
    }
    
    this.vectorType = vectorType;

    this.fullLengthCds = fullLengthCds;

    if (fullLengthCds == null) {
      throw new com.poesys.db.InvalidParametersException("fullLengthCds is required for " + key.getValueList());
    }
    
    this.length = length;

    if (length == null) {
      throw new com.poesys.db.InvalidParametersException("length is required for " + key.getValueList());
    }
    
    this.start = start;

    if (start == null) {
      throw new com.poesys.db.InvalidParametersException("start is required for " + key.getValueList());
    }
    
    this.end = end;

    if (end == null) {
      throw new com.poesys.db.InvalidParametersException("end is required for " + key.getValueList());
    }
    
    // Setter arrays
    querySetters = new java.util.concurrent.CopyOnWriteArrayList<com.poesys.db.dto.ISet>();
    insertQuerySetters = new java.util.concurrent.CopyOnWriteArrayList<com.poesys.db.dto.ISet>();
    insertSetters = new java.util.concurrent.CopyOnWriteArrayList<com.poesys.db.dto.ISet>();
    preSetters = new java.util.concurrent.CopyOnWriteArrayList<com.poesys.db.dto.ISet>();
    postSetters = new java.util.concurrent.CopyOnWriteArrayList<com.poesys.db.dto.ISet>();
    readObjectSetters = new java.util.concurrent.CopyOnWriteArrayList<com.poesys.db.dto.ISet>();
    
    // Add the setters for the geneModel property.
    querySetters.add(new QueryGeneModelSetter());
    
    // Set the internal key data member for the geneModel object retrieval.
    geneModelKey = org.tair.db.locusdetail.LocusdetailFactory.getLocusGeneModelPrimaryKey(geneModelName/* Source: AddExplicitSubKeyProperties + addNaturalSubkeyOnClass + addAssociationForeignKeys: geneModelName Key: true */, locusTairObjectId/* Source: AddNaturalKeyProperties + AddParentKeyAttributes + addAssociationForeignKeys: locusTairObjectId Key: true */);
    // Add a setter to instantiate the required geneModel object before insert.
    insertQuerySetters.add(new QueryGeneModelSetter());
    
    // Add the many-to-many collection setters for the aliases property.
    querySetters.add(new QueryAliasesSetter());
    readObjectSetters.add(new ReadAliasesSetter());
  }

  @Override
  public boolean equals(Object arg0) {
    AbstractLocusGeneModelCdna other = (AbstractLocusGeneModelCdna)arg0;
    return other.key.equals(key);
  }

  @Override
  public int hashCode() {
    return key.hashCode();
  }

  @Override
  public int compareTo(com.poesys.db.dto.IDbDto o) {
    AbstractLocusGeneModelCdna other = (AbstractLocusGeneModelCdna)o;
    // Sort on the key. Same semantics as equals and hashCode().
    return other.key.compareTo(key);
  }
  
  @Override
  public String getSubsystem() {
    return "org.tair.db.locusdetail";
  }

  @Override
  public void markChildrenDeleted() throws DtoStatusException {
    // Only process deletes if child exists.
    if (aliases != null) {
      for (com.poesys.db.dto.IDbDto dto : aliases) {
        dto.cascadeDelete();
      }
    }
  }

  @Override
  public IPrimaryKey getPrimaryKey() {
    return key;
  }
  
  @Override
  public java.sql.Connection getConnection() throws java.sql.SQLException {
    java.sql.Connection connection = null;
    
	/* 
	 * The resource bundle for the DTO's subsystem contains the suffix that
	 * distinguishes multiple versions of the subsystem in the database.properties
	 * file, such as "prod" or "test". Most database.properties files have only
	 * one implementation and use external facilities to switch between the
	 * databases (JNDI, for example, or producing different database.properties
	 * files in different setups). Use the subsystem resource bundle to get the
	 * suffix, then use the full subsystem name to get a connection factory for
	 * the DTO's subsystem, then use that factory to get a JDBC connection.
	 */
	java.util.ResourceBundle rb = 
	  java.util.ResourceBundle.getBundle("org.tair.db.locusdetail");
    String subsystem =
      "org.tair.db.locusdetail"
          + (rb.getString("suffix") == null
             || rb.getString("suffix").length() == 0 ? ""
              : "." + rb.getString("suffix"));
	  
	try {
	  connection = 
	    com.poesys.db.connection.ConnectionFactoryFactory.getInstance(subsystem).getConnection();
	} catch (com.poesys.db.InvalidParametersException e) {
	  throw new java.sql.SQLException(e.getMessage());
	} catch (java.io.IOException e) {
	  throw new java.sql.SQLException(e.getMessage());
	}
	
	return connection;
  }

  /**
   * Nested property geneModelName
   *
   * <p>
   * Composite super-key attribute that uniquely identifies child combined with child sub-key and any other parent super-keys
   * </p>
   *
   * Added by AddExplicitSubKeyProperties + addNaturalSubkeyOnClass + AddParentKeyAttributes
   * Class is read/write: false
   * Class is immutable: false
   * Property is read/write: false
   * Property is lazy: false
   */
  private java.lang.String geneModelName;
  

  public java.lang.String getGeneModelName() {
    return geneModelName;
  }

  /**
   * Clear the geneModelName data member; override in proxy if lazily loaded,
   * otherwise this method does nothing.
   */
  public void clearGeneModelName() {
    // Override in proxy if lazily loaded; otherwise does nothing
  }

  /**
   * Nested property locusTairObjectId
   *
   * <p>
   * Composite super-key attribute that uniquely identifies child combined with child sub-key and any other parent super-keys
   * </p>
   *
   * Added by AddNaturalKeyProperties + AddParentKeyAttributes + AddParentKeyAttributes
   * Class is read/write: false
   * Class is immutable: false
   * Property is read/write: false
   * Property is lazy: false
   */
  private java.lang.Long locusTairObjectId;
  

  public java.lang.Long getLocusTairObjectId() {
    return locusTairObjectId;
  }

  /**
   * Clear the locusTairObjectId data member; override in proxy if lazily loaded,
   * otherwise this method does nothing.
   */
  public void clearLocusTairObjectId() {
    // Override in proxy if lazily loaded; otherwise does nothing
  }

  /**
   * Nested property sequenceId
   *
   * <p>
   * the unique identifier for the cDNA
   * </p>
   *
   * Added by AddExplicitSubKeyProperties + addNaturalSubkeyOnClass
   * Class is read/write: false
   * Class is immutable: false
   * Property is read/write: false
   * Property is lazy: false
   */
  private java.lang.Long sequenceId;
  

  public java.lang.Long getSequenceId() {
    return sequenceId;
  }

  /**
   * Clear the sequenceId data member; override in proxy if lazily loaded,
   * otherwise this method does nothing.
   */
  public void clearSequenceId() {
    // Override in proxy if lazily loaded; otherwise does nothing
  }

  /**
   * Nested property sequenceName
   *
   * <p>
   * the sequence name
   * </p>
   *
   * Added by AddLocalAttributeProperties
   * Class is read/write: false
   * Class is immutable: false
   * Property is read/write: false
   * Property is lazy: false
   */
  private java.lang.String sequenceName;
  

  public java.lang.String getSequenceName() {
    return sequenceName;
  }

  /**
   * Clear the sequenceName data member; override in proxy if lazily loaded,
   * otherwise this method does nothing.
   */
  public void clearSequenceName() {
    // Override in proxy if lazily loaded; otherwise does nothing
  }

  /**
   * Nested property icAccession
   *
   * <p>
   * the unique identifier at any of the major international genomics data centers
   * for the cDNA sequence
   * </p>
   *
   * Added by AddLocalAttributeProperties
   * Class is read/write: false
   * Class is immutable: false
   * Property is read/write: false
   * Property is lazy: false
   */
  private java.lang.String icAccession;
  

  public java.lang.String getIcAccession() {
    return icAccession;
  }

  /**
   * Clear the icAccession data member; override in proxy if lazily loaded,
   * otherwise this method does nothing.
   */
  public void clearIcAccession() {
    // Override in proxy if lazily loaded; otherwise does nothing
  }

  /**
   * Nested property vectorType
   *
   * <p>
   * the kind of vector associated with the cDNA:
   * </p>
   * <p>
   * BAC
   * </p>
   * <p>
   * BiBAC
   * </p>
   * <p>
   * P1
   * </p>
   * <p>
   * TAC
   * </p>
   * <p>
   * YAC
   * </p>
   * <p>
   * cosmid
   * </p>
   * <p>
   * lambda
   * </p>
   * <p>
   * plasmid
   * </p>
   * <p>
   * unknown
   * </p>
   *
   * Added by AddLocalAttributeProperties
   * Class is read/write: false
   * Class is immutable: false
   * Property is read/write: false
   * Property is lazy: false
   */
  private java.lang.String vectorType;
  

  public java.lang.String getVectorType() {
    return vectorType;
  }

  /**
   * Clear the vectorType data member; override in proxy if lazily loaded,
   * otherwise this method does nothing.
   */
  public void clearVectorType() {
    // Override in proxy if lazily loaded; otherwise does nothing
  }

  /**
   * Nested property fullLengthCds
   *
   * <p>
   * whether this cDNA is a "full-length" cDNA that contains the entire coding region
   * of the gene
   * </p>
   *
   * Added by AddLocalAttributeProperties
   * Class is read/write: false
   * Class is immutable: false
   * Property is read/write: false
   * Property is lazy: false
   */
  private java.lang.Boolean fullLengthCds;
  

  public java.lang.Boolean getFullLengthCds() {
    return fullLengthCds;
  }

  /**
   * Clear the fullLengthCds data member; override in proxy if lazily loaded,
   * otherwise this method does nothing.
   */
  public void clearFullLengthCds() {
    // Override in proxy if lazily loaded; otherwise does nothing
  }

  /**
   * Nested property length
   *
   * <p>
   * the length of the cDNA sequence, unit is bp (base pairs)
   * </p>
   *
   * Added by AddLocalAttributeProperties
   * Class is read/write: false
   * Class is immutable: false
   * Property is read/write: false
   * Property is lazy: false
   */
  private java.lang.Long length;
  

  public java.lang.Long getLength() {
    return length;
  }

  /**
   * Clear the length data member; override in proxy if lazily loaded,
   * otherwise this method does nothing.
   */
  public void clearLength() {
    // Override in proxy if lazily loaded; otherwise does nothing
  }

  /**
   * Nested property start
   *
   * <p>
   * the starting position on the chromosome in unit bp (base pairs)
   * </p>
   *
   * Added by AddLocalAttributeProperties
   * Class is read/write: false
   * Class is immutable: false
   * Property is read/write: false
   * Property is lazy: false
   */
  private java.lang.Long start;
  

  public java.lang.Long getStart() {
    return start;
  }

  /**
   * Clear the start data member; override in proxy if lazily loaded,
   * otherwise this method does nothing.
   */
  public void clearStart() {
    // Override in proxy if lazily loaded; otherwise does nothing
  }

  /**
   * Nested property end
   *
   * <p>
   * the ending position on the chromosome in unit bp (base pairs)
   * </p>
   *
   * Added by AddLocalAttributeProperties
   * Class is read/write: false
   * Class is immutable: false
   * Property is read/write: false
   * Property is lazy: false
   */
  private java.lang.Long end;
  

  public java.lang.Long getEnd() {
    return end;
  }

  /**
   * Clear the end data member; override in proxy if lazily loaded,
   * otherwise this method does nothing.
   */
  public void clearEnd() {
    // Override in proxy if lazily loaded; otherwise does nothing
  }

  /**
   * Nested property geneModel
   *
   * <p>
   * the gene model to which the cDNA belongs
   * </p>
   *
   * Added by AddToOneAssociationRequiredObjectProperties
   * Class is read/write: false
   * Class is immutable: false
   * Property is read/write: false
   * Property is lazy: false
   */
  private org.tair.db.locusdetail.ILocusGeneModel geneModel;
  

  public org.tair.db.locusdetail.ILocusGeneModel getGeneModel() {
    return geneModel;
  }

  /**
   * Clear the geneModel data member; override in proxy if lazily loaded,
   * otherwise this method does nothing.
   */
  public void clearGeneModel() {
    // Override in proxy if lazily loaded; otherwise does nothing
  }

  /**
   * Nested property aliases
   *
   * <p>
   * the set of alternative names for the cDNA sequence
   * </p>
   *
   * Added by AddToManyChildCollectionProperties
   * Class is read/write: false
   * Class is immutable: false
   * Property is read/write: false
   * Property is lazy: false
   */
  // Doesn't serialize; package access allows proxy to set on readObject()
  transient java.util.Collection<org.tair.db.locusdetail.ILocusGeneModelCdnaAlias> aliases;
  // Ordered list of keys of the objects in the $property.uniquename list
  java.util.List<com.poesys.db.pk.IPrimaryKey> aliasesKeys = 
    new java.util.ArrayList<com.poesys.db.pk.IPrimaryKey>();
  

  public java.util.Collection<org.tair.db.locusdetail.ILocusGeneModelCdnaAlias> getAliases() {
    return aliases;
  }

  /**
   * Clear the aliases data member; override in proxy if lazily loaded,
   * otherwise this method does nothing.
   */
  public void clearAliases() {
    // Override in proxy if lazily loaded; otherwise does nothing
  }

  @Override
  public void update(com.poesys.db.dto.ISubject subject,
                     com.poesys.db.dao.DataEvent event)
      throws com.poesys.db.dto.DtoStatusException {

    // Clean up aliases.
    if (subject != null && subject instanceof org.tair.db.locusdetail.ILocusGeneModelCdnaAlias && 
        event == com.poesys.db.dao.DataEvent.DELETE &&
        aliases != null) {
      // Delete to-many aliases child from collection
      aliases.remove(subject);
      subject.detach(this, com.poesys.db.dao.DataEvent.MARKED_DELETED);
      subject.detach(this, com.poesys.db.dao.DataEvent.DELETE);
    }
  }
}