/**
 * Copyright 2012 Carnegie Institution for Science. All rights reserved.
 */
// Generated by AndroMDA Poesys/DB cartridge: DO NOT EDIT DIRECTLY.
// Template: AbstractDelegateTest.vsl

package org.tair.bs.genomicregion;


import static org.junit.Assert.assertTrue;
import org.junit.After;

import com.poesys.db.connection.ConnectionException;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

import org.apache.log4j.Logger;
import org.junit.Before;
import org.junit.Test;

import com.poesys.bs.delegate.DelegateException;
import com.poesys.db.InvalidParametersException;
import com.poesys.db.dao.DaoManagerFactory;
import com.poesys.db.dao.IDaoManager;


/**
 * <p>
 * Test the TranscriptDelegate class.
 * </p>
 * <p>
 * Note that all the test cases defined here run independently, because JUnit
 * does not guarantee the execution order of the test cases in any way.
 * </p>
 * 
 * @author Poesys/DB Cartridge
 */
public abstract class AbstractTranscriptDelegateTest {
  /** Define a class logger. */
  protected static Logger logger = Logger.getLogger(AbstractTranscriptDelegateTest.class);
  /** Subsystem name in database.properties file */
  protected String subsystem = "org.tair.db.genomicregion";
  /** Boolean saved off for later comparison in update tests */
  protected Boolean originalBooleanValue = null;
  /** Delegate created at class level for sharing between methods */
  protected TranscriptDelegate delegate = null;

  /**
   * Set up the test by creating the class under test (CUT).
   * 
   * @throws InvalidParametersException when an invalid parameter prevents the
   *       creation of an object
   * @throws DelegateException when a problem occurs during object creation
   */
  @Before
  public void setUp() {
    delegate = GenomicregionDelegateFactory.getTranscriptDelegate();
    // Clear any existing caches of objects using the memory manager set in
    // the creation of the delegate.
    clearCaches();
    assertTrue("No delegate created", delegate != null);
  }
  
  /**
   * Helper method to clear caches for any objects used in the test
   */
  protected void clearCaches() {
    // Primary class Transcript cache
    IDaoManager manager = DaoManagerFactory.getManager(getSubsystem());
    if (manager != null) {
      manager.clearCache(org.tair.db.genomicregion.Transcript.class.getName());
    }

    // Clear nested object caches
    
    // Child Est cache
    IDaoManager EstManager = DaoManagerFactory.getManager("org.tair.db.genomicregion");
    if (EstManager != null) {
      EstManager.clearCache(org.tair.db.genomicregion.Est.class.getName());
    }
    
    // Child Polypeptide cache
    IDaoManager PolypeptideManager = DaoManagerFactory.getManager("org.tair.db.genomicregion");
    if (PolypeptideManager != null) {
      PolypeptideManager.clearCache(org.tair.db.genomicregion.Polypeptide.class.getName());
    }
    
    // Child ExperimentalCdna cache
    IDaoManager ExperimentalCdnaManager = DaoManagerFactory.getManager("org.tair.db.genomicregion");
    if (ExperimentalCdnaManager != null) {
      ExperimentalCdnaManager.clearCache(org.tair.db.genomicregion.ExperimentalCdna.class.getName());
    }
  }

  /**
   * Create some number of new Transcript objects and return them in a list.  
   * This is a helper method for writable children DTOs.
   * 
   * @param count the number of objects to create
   * @return the stored object
   * @throws InvalidParametersException when an invalid parameter prevents the
   *       creation of an object
   * @throws DelegateException when a problem occurs during object creation
   */
  protected List<org.tair.bs.genomicregion.BsTranscript> createGenomicregionTranscript(int count) 
      throws DelegateException, InvalidParametersException {
    List<org.tair.bs.genomicregion.BsTranscript> objects = 
      new CopyOnWriteArrayList<org.tair.bs.genomicregion.BsTranscript>();
    @SuppressWarnings("unused")
    java.util.Random r = new java.util.Random();

    // Create the required TranscribesGenomicregionGene objects.
    List<org.tair.bs.genomicregion.BsGene> transcribesGenomicregionGeneObjects = createTranscribesGenomicregionGene(count);    
    
    for (int i = 0; i < count; i++) {
      java.lang.Long transcribesTairObjectId = transcribesGenomicregionGeneObjects.get(i).getTairObjectId();
      java.lang.Long tairObjectId = r.nextLong();
      java.lang.String name = 
        com.poesys.cartridges.db.utilities.StringUtilities.generateString(50);
      // Use random integer here because long might be out of range.
      java.sql.Timestamp dateEntered = new java.sql.Timestamp(r.nextInt(999999999));
      java.lang.Long parentRegionTairObjectId = r.nextLong();
      java.lang.String genome = 
        com.poesys.cartridges.db.utilities.StringUtilities.generateString();
      java.lang.String sequence = 
        com.poesys.cartridges.db.utilities.StringUtilities.generateString();
      java.lang.Long seqTairObjectId = r.nextLong();
      java.lang.String genbankAccession = 
        com.poesys.cartridges.db.utilities.StringUtilities.generateString(20);
      java.lang.String primarySymbol = 
        com.poesys.cartridges.db.utilities.StringUtilities.generateString(50);
      java.lang.String primarySymbolFullName = 
        com.poesys.cartridges.db.utilities.StringUtilities.generateString(200);
      java.lang.Long startPosition = r.nextLong();
      java.lang.Long endPosition = r.nextLong();
      java.lang.Long length = r.nextLong();
      java.lang.String orientation = 
        com.poesys.cartridges.db.utilities.StringUtilities.generateString();
      java.lang.String basis = 
        com.poesys.cartridges.db.utilities.StringUtilities.generateString();
      java.lang.Boolean derived = r.nextBoolean();
      java.lang.Boolean representativeGeneModel = r.nextBoolean();

      // Create the object.
      BsTranscript object = null;
      try {
        object = delegate.createTranscript(tairObjectId, name, dateEntered, parentRegionTairObjectId, genome, sequence, seqTairObjectId, genbankAccession, primarySymbol, primarySymbolFullName, startPosition, endPosition, length, orientation, basis, derived, representativeGeneModel, transcribesTairObjectId);
        // Add any required nested Gene object.
        object.setTranscribes(transcribesGenomicregionGeneObjects.get(i));
      } catch (InvalidParametersException e) {
        Object[] args = e.getParameters().toArray();
        String message = com.poesys.db.Message.getMessage(e.getMessage(), args);
        throw new RuntimeException(message, e);
      } catch (DelegateException e) {
        throw new RuntimeException(e.getMessage(), e);
      }
        
      objects.add(object);
    }
    
    return objects;
  }

  /**
   * <p>
   * Create some number of new Polypeptide objects and return them in a list.  
   * This is a helper method that the concrete subclass needs to implement. It
   * returns a child object.
   * </p>
   * <ul>
   * <li>Create the list to return as a CopyOnWriteArrayList&lt;BsPolypeptide&gt;</li>
   * <li>Generate any required objects linked to this object other than the 
   * parent by calling the appropriate delegate's create method</li>
   * <li>Create the sub-key as an integer for ordered subkeys or a random string
   * for unordered subkeys</li>
   * <li>Generate random values for any primitive data members</li>
   * <li>Create the BsPolypeptide by calling delegate.createBsPolypeptide</li>
   * <li>Add the new child to the list</li>
   * </ul>
   * <p>
   * Key type: CompositeKey
   * </p>
   * @param parent the parent Transcript
   * @param count the number of objects to create
   * @return the stored object
   * @throws DelegateException when there is a problem creating an object
   * @throws InvalidParametersException when there is some problem with the 
   *                                    input parameters for creating an object
   */
  abstract protected List<org.tair.bs.genomicregion.BsPolypeptide> createGenomicregionPolypeptide(org.tair.db.genomicregion.ITranscript parent, 
                                                       int count) 
      throws DelegateException, InvalidParametersException; // create 2
      
  /**
   * <p>
   * Create some number of new Gene objects and return them in a list.  
   * This is a helper method that the concrete subclass needs to implement. It
   * returns the "child" corresponding to the superclass.
   * </p>
   * <ul>
   * <li>Create the list to return as a CopyOnWriteArrayList&lt;BsGene&gt;</li>
   * <li>Generate random values for any primitive data members</li>
   * <li>Create the BsGene by calling delegate.createBs${foreignDto.name}</li>
   * <li>Add the new child to the list</li>
   * </ul>
   * <p>
   * Key type: None
   * </p>
   * @param count the number of objects to create
   * @return the stored object
   * @throws DelegateException when there is a problem creating an object
   * @throws InvalidParametersException when there is some problem with the 
   *                                    input parameters for creating an object
   */
  abstract protected List<org.tair.bs.genomicregion.BsGene> createTranscribesGenomicregionGene(int count) 
      throws DelegateException, InvalidParametersException; // create 4
      
  /**
   * Test method for delegate insert
   */
  @Test
  public void testInsert() {
    // Create a new Transcript object to perform the test.
    List<BsTranscript> objects = createGenomicregionTranscript(1);
    assertTrue("No object created", objects.get(0) != null);
    delegate.insert(objects);
    com.poesys.db.pk.NaturalPrimaryKey key = 
      (com.poesys.db.pk.NaturalPrimaryKey)objects.get(0).getPrimaryKey();
    assertTrue("No key for inserted object", key != null);

    BsTranscript queriedObject = queryStoredObject(key);
    assertTrue("No queried inserted object", queriedObject != null);
    assertTrue("Wrong object", objects.get(0).equals(queriedObject));
    // Delete the inserted objects to clean up.
    // Mark all the objects for delete.
    for (BsTranscript object : objects) {
      object.delete();
    }

    delegate.deleteBatch(objects);
  }

  /**
   * Test method for getObject
   * 
   * @throws java.sql.SQLException when there is a problem querying a collection 
   *                               property
   */
  @Test
  public void testGetObject() throws java.sql.SQLException {
    // Create a new Transcript object to perform the test.
    List<BsTranscript> objects = createGenomicregionTranscript(1);
    assertTrue("No object created", objects.get(0) != null);
    delegate.insert(objects);
    com.poesys.db.pk.NaturalPrimaryKey key = 
      (com.poesys.db.pk.NaturalPrimaryKey)objects.get(0).getPrimaryKey();
    assertTrue("No key generated from concrete implementation", key != null);
    BsTranscript insertedObject = objects.get(0);
    assertTrue("No comparison object for object query", insertedObject != null);
    
    // Query the object.
    BsTranscript object = delegate.getObject(key);
    assertTrue("Couldn't get object", object != null);
    assertTrue("Wrong object", insertedObject.equals(object));
    // Delete the inserted objects to clean up.
    // Mark all the objects for delete.
    for (BsTranscript o : objects) {
      o.delete();
    }

    delegate.deleteBatch(objects);
  }

  /**
   * <p>
   * Test method for getAllObjects
   * </p>
   * <p>
   * For this read/write class, the method ensures that there are objects to
   * query by inserting an object, querying, then deleting the object.
   * </p>
   *
   * @throws InvalidParametersException when an invalid parameter prevents the
   *       creation of an object
   * @throws DelegateException when a problem occurs during object creation
   */
  @Test
  public void testGetAllObjects() 
       throws InvalidParametersException, DelegateException {
    // Create a new Transcript object to perform the test.
    List<BsTranscript> insertedObjects = createGenomicregionTranscript(1);
    assertTrue("No object created", insertedObjects.get(0) != null);
    delegate.insert(insertedObjects);
    
    // Query all the objects.
    List<BsTranscript> objects = delegate.getAllObjects(100);
    assertTrue("Couldn't create list", objects != null);
    // The list should not be empty.
    assertTrue("List of all objects empty", objects.size() != 0);
    // Delete the inserted objects to clean up.
    // Mark all the objects for delete.
    for (BsTranscript o : insertedObjects) {
      o.delete();
    }

    delegate.deleteBatch(insertedObjects);
  }

  /**
   * A helper method that queried a previously stored object identified by its
   * primary key. The method returns the queried object for further testing
   * after performing structural tests on the object.
   * 
   * @param key the primary key of the object to retrieve
   * @return the object
   * @throws DelegateException when there is a problem querying the object
   */
  protected BsTranscript queryStoredObject(com.poesys.db.pk.NaturalPrimaryKey key)
      throws DelegateException {
    // Clear the caches, then query the saved object and test it.
    clearCaches();

    BsTranscript queriedObject = delegate.getObject(key);
    assertTrue("Object not found", queriedObject != null);
    return queriedObject;
  }

  /**
   * A helper method that tests the stored existence of the object identified
   * by the pre-existing, saved primary key.
   * 
   * @param key the primary key (com.poesys.db.pk.NaturalPrimaryKey) of the object to check
   * @return true if the object is in the database, false if not
   * @throws DelegateException when there is a problem querying the object
   */
  protected boolean exists(com.poesys.db.pk.NaturalPrimaryKey key) throws DelegateException {
    // Clear the caches, then query the saved object and test it.
    clearCaches();

    BsTranscript queriedObject = delegate.getObject(key);
    return queriedObject != null;
  }

  /**
   * Test the updating of the inserted object.
   */
  @Test
  public void testUpdate() {
    // Create a new Transcript object to perform the test.
    List<BsTranscript> objects = createGenomicregionTranscript(1);
    assertTrue("No object created", objects.get(0) != null);
    delegate.insert(objects);
    com.poesys.db.pk.NaturalPrimaryKey key = 
      (com.poesys.db.pk.NaturalPrimaryKey)objects.get(0).getPrimaryKey();
    assertTrue("No key for inserted object", key != null);
    BsTranscript insertedObject = objects.get(0);
    assertTrue("No created object to update", insertedObject != null);
    
    // Update the inserted object and query it to test.
    updateColumn(insertedObject);
    delegate.update(insertedObject);
    BsTranscript queriedObject = queryStoredObject(key);
    assertTrue("Object not updated", isColumnUpdated(queriedObject));
    // Delete the inserted objects to clean up.
    // Mark all the objects for delete.
    for (BsTranscript object : objects) {
      object.delete();
    }

    delegate.deleteBatch(objects);
  }
  
  /**
   * Update the first mutable column with a new value.
   *
   * @param object the BsTranscript object to update
   */
  protected void updateColumn(BsTranscript object) {
    try {
      object.setSequence("Transcript.sequence updated");
    } catch (com.poesys.db.dto.DtoStatusException e) {
      org.junit.Assert.fail(e.getMessage());
    }
  }

  /**
   * Is the appropriate column in the input object set to the update value?
   *
   * @param object the BsTranscript object to update
   * @return true if the column is updated, false if not
   * @see #updateColumn
   */
  protected boolean isColumnUpdated(BsTranscript object) {
    boolean retVal = false;
    retVal = object.getSequence().equals("Transcript.sequence updated");
    return retVal;
  }

  /**
   * Is the appropriate column in the input object set to the updated value in 
   * the original object? This method is a helper for the batchUpdate test.
   *
   * @param original the original, updated BsTranscript object
   * @param queried the queried BsTranscript object
   * @return true if the column is updated, false if not
   */
  protected boolean isColumnUpdated(BsTranscript original, BsTranscript queried) {
    boolean retVal = true;
    retVal = 
      retVal && queried.getSequence().equals(original.getSequence());
    retVal = 
      retVal && queried.getSeqTairObjectId().equals(original.getSeqTairObjectId());
    retVal = 
      retVal && queried.getGenbankAccession().equals(original.getGenbankAccession());
    retVal = 
      retVal && queried.getPrimarySymbol().equals(original.getPrimarySymbol());
    retVal = 
      retVal && queried.getPrimarySymbolFullName().equals(original.getPrimarySymbolFullName());
    retVal = 
      retVal && queried.getStartPosition().equals(original.getStartPosition());
    retVal = 
      retVal && queried.getEndPosition().equals(original.getEndPosition());
    retVal = 
      retVal && queried.getLength().equals(original.getLength());
    retVal = 
      retVal && queried.getOrientation().equals(original.getOrientation());
    retVal = 
      retVal && queried.getBasis().equals(original.getBasis());
    retVal = 
      retVal && queried.getDerived().equals(original.getDerived());
    retVal = 
      retVal && queried.getRepresentativeGeneModel().equals(original.getRepresentativeGeneModel());
    return retVal;
  }

  /**
   * Test method for batch update
   * 
   * @throws InvalidParametersException when an invalid parameter prevents the
   *       creation of an object
   * @throws DelegateException when a problem occurs during object creation
   */
  @Test
  public void testUpdateBatch() throws InvalidParametersException,
      DelegateException {
    List<BsTranscript> objects = createGenomicregionTranscript(2);
    delegate.insert(objects);
    // Allocate a map to hold the updated objects for later comparison.
    java.util.Map<com.poesys.db.pk.IPrimaryKey, BsTranscript> map = 
      new java.util.TreeMap<com.poesys.db.pk.IPrimaryKey, BsTranscript>();
    for (BsTranscript object : objects) {
      updateColumn(object);
      // Add the object to the storage map.
      map.put(object.getPrimaryKey(), object);
    }
    delegate.updateBatch(objects);
    
    clearCaches();
    
    for (BsTranscript object : objects) {
      BsTranscript queriedObject = 
        queryStoredObject((com.poesys.db.pk.NaturalPrimaryKey)object.getPrimaryKey());
      BsTranscript originalObject = map.get(object.getPrimaryKey());
      assertTrue("Object not updated", 
                 isColumnUpdated(originalObject, queriedObject));
    }
    // Delete the inserted objects to clean up.
    // Mark all the objects for delete.
    for (BsTranscript object : objects) {
      object.delete();
    }

    delegate.deleteBatch(objects);
  }

  /**
   * Test method for delete
   * 
   * @throws InvalidParametersException when an invalid parameter prevents the
   *       creation of an object
   * @throws DelegateException when a problem occurs during object creation
   */
  @Test
  public void testDelete() throws InvalidParametersException,
      DelegateException {
    // Create a new Transcript object to perform the test.
    List<BsTranscript> objects = createGenomicregionTranscript(1);
    assertTrue("No object created", objects.get(0) != null);
    delegate.insert(objects);
    com.poesys.db.pk.NaturalPrimaryKey key = 
      (com.poesys.db.pk.NaturalPrimaryKey)objects.get(0).getPrimaryKey();
    assertTrue("No key for inserted object", key != null);
    BsTranscript insertedObject = objects.get(0);
    assertTrue("No inserted object to delete", insertedObject != null);
    
    delegate.delete(insertedObject);
    assertTrue("object not deleted", !exists(key));
    // Delete the inserted objects to clean up.
    // Mark all the objects for delete.
    for (BsTranscript object : objects) {
      object.delete();
    }

    delegate.deleteBatch(objects);
  }

  /**
   * Test method for deleteBatch
   * 
   * @throws InvalidParametersException when an invalid parameter prevents the
   *       creation of an object
   * @throws DelegateException when a problem occurs during object creation
   */
  @Test
  public void testDeleteBatch() throws InvalidParametersException,
      DelegateException {
    List<BsTranscript> objects = createGenomicregionTranscript(2);
    delegate.insert(objects);
    
    // Mark all the objects for delete.
    for (BsTranscript object : objects) {
      object.delete();
    }

    delegate.deleteBatch(objects);

    for (BsTranscript object : objects) {
      assertTrue("object not deleted", !exists((com.poesys.db.pk.NaturalPrimaryKey)object.getPrimaryKey()));
    }
  }

  /**
   * Test method for process()
   * 
   * @throws InvalidParametersException when an invalid parameter prevents the
   *       creation of an object
   * @throws DelegateException when a problem occurs during object creation
   */
  @Test
  public void testProcess() throws InvalidParametersException,
      DelegateException {
    // Create 3 rows--one to insert, one to update, one to delete.
    List<BsTranscript> allObjects = createGenomicregionTranscript(3);
    // Insert the first two rows to later update and delete.
    List<BsTranscript> existingObjects = 
      new CopyOnWriteArrayList<BsTranscript>();
    existingObjects.add(allObjects.get(0));
    existingObjects.add(allObjects.get(1));
    delegate.insert(existingObjects);
    updateColumn(existingObjects.get(0));
    existingObjects.get(1).delete();

    // Set the third object as the object to insert.
    List<BsTranscript> insertObject = 
      new CopyOnWriteArrayList<BsTranscript>();
    insertObject.add(allObjects.get(2));
    
    // Put it all together.
    List<BsTranscript> objects = 
      new CopyOnWriteArrayList<BsTranscript>(existingObjects);
    objects.addAll(insertObject);

    // Test the process method
    delegate.process(objects);

    // Verify the update
    BsTranscript queriedObject = 
      queryStoredObject((com.poesys.db.pk.NaturalPrimaryKey)existingObjects.get(0).getPrimaryKey());
    assertTrue("Update not processed", isColumnUpdated(queriedObject));

    // Verify the delete
    assertTrue("object not deleted", !exists((com.poesys.db.pk.NaturalPrimaryKey)existingObjects.get(1).getPrimaryKey()));

    // Verify the insert
    assertTrue("object not inserted", exists((com.poesys.db.pk.NaturalPrimaryKey)insertObject.get(0).getPrimaryKey()));
    
    // Delete the inserted objects to clean up.
    // Mark all the objects for delete.
    for (BsTranscript object : allObjects) {
      object.delete();
    }

    delegate.deleteBatch(allObjects);
  }

  /**
   * Test the creation and adding of a child BsPolypeptide object to the BsTranscript
   * as a proteins link.
   */
  @Test
  public void testUpdateAddProteinsPolypeptide() {
    // Create a new BsTranscript object to perform the test.
    List<BsTranscript> objects = createGenomicregionTranscript(1);
    assertTrue("No object created", objects.get(0) != null);
    delegate.insert(objects);
    com.poesys.db.pk.NaturalPrimaryKey key = 
      (com.poesys.db.pk.NaturalPrimaryKey)objects.get(0).getPrimaryKey();
    assertTrue("No key for inserted object", key != null);

    // Test to make sure the object is there.
    BsTranscript queriedObject = queryStoredObject(key);
    assertTrue("No queried inserted object", queriedObject != null);
    assertTrue("Wrong object", objects.get(0).equals(queriedObject));

    // Add a child and update the parent to insert the child.
    List<BsPolypeptide> children = createGenomicregionPolypeptide(queriedObject.toDto(), 1);
    queriedObject.addProteinsPolypeptide(children.get(0));
    delegate.update(queriedObject);
    
    // Test to make sure the object is there.
    BsTranscript queriedUpdatedObject = queryStoredObject(key);
    assertTrue("No queried updated object", queriedUpdatedObject != null);
    assertTrue("Wrong object", objects.get(0).equals(queriedUpdatedObject));
    assertTrue("No proteins, Polypeptide not inserted", 
               queriedUpdatedObject.getProteins().size() == 1);
    // Delete the inserted objects to clean up.
    // Mark all the objects for delete.
    for (BsTranscript object : objects) {
      object.delete();
    }

    delegate.deleteBatch(objects);
  }

  /**
   * Test method for truncateTable
   */
  @Test
  public void testTruncateTable() {
    delegate.truncateTable("Transcript");
    List<BsTranscript> objects = delegate.getAllObjects(100);
    assertTrue("Couldn't create list", objects != null);
    // The list should be empty.
    assertTrue("Table not truncated", objects.size() == 0);
  }

  /**
   * Return the database access subsystem name to use to look up access
   * properties in the database.properties file. The concrete implementation
   * should override this method to provide a different subsystem name if
   * needed.
   * 
   * @return the database access subsystem name
   */
  protected String getSubsystem() {
    return subsystem;
  }
  
  /**
   * Flush all resources associated with connections.
   * 
   * @throws ConnectionException when there is a problem flushing resources
   */
  @After
  public void tearDown() throws ConnectionException {
    delegate.flush();
  }
}
