/**
 * Copyright 2012 Carnegie Institution for Science. All rights reserved.
 */
// Generated by AndroMDA Poesys/DB cartridge: DO NOT EDIT DIRECTLY.
// Template: AbstractBsDto.vsl


package org.tair.bs.genomicregion;


import com.poesys.bs.delegate.DelegateException;
import com.poesys.bs.dto.IDto;
import com.poesys.db.dto.DtoStatusException;
import com.poesys.db.pk.IPrimaryKey;
import com.poesys.bs.dto.AbstractDto;


import org.tair.db.genomicregion.IExperimentalCdna;
import org.tair.db.genomicregion.ExperimentalCdnaProxy;
import org.tair.db.genomicregion.ExperimentalCdna;


/**
 * <p>
 * A business layer data-transfer object (DTO) for the ExperimentalCdna. This class
 * is an abstract class that contains AndroMDA generated code; change nothing
 * in this class. Instead, override any methods in the concrete subclass
 * generated in the same package. AndroMDA will overwrite this class each time
 * you run it but will never overwrite the concrete subclass.
 * </p>
 * <p>
 * SO_0000756: DNA synthesized by reverse transcriptase using RNA as a template,
 * produced by in vivo experimentation; based on clone
 * </p>
 * <p>
 * Stereotypes:
 * </p>
 * <ul>
 *     <li>CompositeKey</li>
 *     <li>Persistent</li>
 * </ul>
 * <p>
 * Class is read/write: false
 * </p>
 *
 * @author Poesys/DB Cartridge
 */
public abstract class AbstractBsExperimentalCdna 
    extends AbstractDto<org.tair.db.genomicregion.IExperimentalCdna> {

  /**
   * Create a BsExperimentalCdna object from a ExperimentalCdna object.
   * 
   * @param dto the data-access layer ExperimentalCdna DTO
   * @throws DelegateException when there is a problem creating the ExperimentalCdna
   */
  public AbstractBsExperimentalCdna(IExperimentalCdna dto) throws DelegateException {
    super(dto);
  }

  /**
   * Create a ExperimentalCdna from new data.
   *
   * @param key the primary key of the ExperimentalCdna
   * @param tairObjectId composite super-key attribute that uniquely identifies child combined with child sub-key and any other parent super-keys
   * @param seqTairObjectId the unique TAIR identifier for the cDNA sequence
   * @param name the sequence name
   * @param sequence the DNA sequence for the cDNA
   * @param genbankAccession the unique identifier at GenBank for the cDNA sequence
   * @param vectorType the kind of vector associated with the cDNA:
BAC
BiBAC
P1
TAC
YAC
cosmid
lambda
plasmid
unknown
   * @param fullLengthCds whether this cDNA is a "full-length" cDNA that contains the entire coding region
of the gene
   * @param length the length of the cDNA sequence, unit is bp (base pairs)
   * @param startPosition the starting position on the chromosome in unit bp (base pairs)
   * @param endPosition the ending position on the chromosome in unit bp (base pairs)
   * @param dateEntered the date on which the protein record was created in the source database
   */
  public AbstractBsExperimentalCdna(IPrimaryKey key, java.lang.Long tairObjectId, java.lang.Long seqTairObjectId, java.lang.String name, java.lang.String sequence, java.lang.String genbankAccession, java.lang.String vectorType, java.lang.Boolean fullLengthCds, java.lang.Long length, java.lang.Long startPosition, java.lang.Long endPosition, java.sql.Timestamp dateEntered) {
    super(new ExperimentalCdnaProxy(new ExperimentalCdna(key, tairObjectId, seqTairObjectId, name, sequence, genbankAccession, vectorType, fullLengthCds, length, startPosition, endPosition, dateEntered)));
  }

  @SuppressWarnings("unchecked")
  @Override
  public boolean equals(Object arg0) {
    // Unchecked cast here
    IDto<IExperimentalCdna> other = (IDto<IExperimentalCdna>)arg0;
    return dto.equals(other.toDto());
  }

  @Override
  public int hashCode() {
    return dto.hashCode();
  }

  public int compareTo(IDto<IExperimentalCdna> o) {
    return dto.compareTo(o.toDto());
  }

  public void markChildrenDeleted() throws DtoStatusException {
    dto.markChildrenDeleted();
  }
  
  public IPrimaryKey getPrimaryKey() {
    return dto.getPrimaryKey();
  }

  // Data member properties

  /**
   * <p>
   * Composite super-key attribute that uniquely identifies child combined with child sub-key and any other parent super-keys
   * </p>
   * <p>
   * Added by AddNaturalKeyProperties + AddInheritedKeyProperties + AddParentKeyAttributes as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: false</li>
   * <li>Property is defined in the data-access object ExperimentalCdna</li>
   * </ul>
   * </p>
   * @return a java.lang.Long tairObjectId
   */
  public java.lang.Long getTairObjectId() {
    return dto.getTairObjectId();
  }

  /**
   * <p>
   * the unique TAIR identifier for the cDNA sequence
   * </p>
   * <p>
   * Added by AddExplicitSubKeyProperties + addNaturalSubkeyOnClass as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: false</li>
   * <li>Property is defined in the data-access object ExperimentalCdna</li>
   * </ul>
   * </p>
   * @return a java.lang.Long seqTairObjectId
   */
  public java.lang.Long getSeqTairObjectId() {
    return dto.getSeqTairObjectId();
  }

  /**
   * <p>
   * the sequence name
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: false</li>
   * <li>Property is defined in the data-access object ExperimentalCdna</li>
   * </ul>
   * </p>
   * @return a java.lang.String name
   */
  public java.lang.String getName() {
    return dto.getName();
  }

  /**
   * <p>
   * the DNA sequence for the cDNA
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: false</li>
   * <li>Property is defined in the data-access object ExperimentalCdna</li>
   * </ul>
   * </p>
   * @return a java.lang.String sequence
   */
  public java.lang.String getSequence() {
    return dto.getSequence();
  }

  /**
   * <p>
   * the unique identifier at GenBank for the cDNA sequence
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: false</li>
   * <li>Property is defined in the data-access object ExperimentalCdna</li>
   * </ul>
   * </p>
   * @return a java.lang.String genbankAccession
   */
  public java.lang.String getGenbankAccession() {
    return dto.getGenbankAccession();
  }

  /**
   * <p>
   * the kind of vector associated with the cDNA:
   * </p>
   * <p>
   * BAC
   * </p>
   * <p>
   * BiBAC
   * </p>
   * <p>
   * P1
   * </p>
   * <p>
   * TAC
   * </p>
   * <p>
   * YAC
   * </p>
   * <p>
   * cosmid
   * </p>
   * <p>
   * lambda
   * </p>
   * <p>
   * plasmid
   * </p>
   * <p>
   * unknown
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: false</li>
   * <li>Property is defined in the data-access object ExperimentalCdna</li>
   * </ul>
   * </p>
   * @return a java.lang.String vectorType
   */
  public java.lang.String getVectorType() {
    return dto.getVectorType();
  }

  /**
   * <p>
   * whether this cDNA is a "full-length" cDNA that contains the entire coding region
   * of the gene
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: false</li>
   * <li>Property is defined in the data-access object ExperimentalCdna</li>
   * </ul>
   * </p>
   * @return a java.lang.Boolean fullLengthCds
   */
  public java.lang.Boolean getFullLengthCds() {
    return dto.getFullLengthCds();
  }

  /**
   * <p>
   * the length of the cDNA sequence, unit is bp (base pairs)
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: false</li>
   * <li>Property is defined in the data-access object ExperimentalCdna</li>
   * </ul>
   * </p>
   * @return a java.lang.Long length
   */
  public java.lang.Long getLength() {
    return dto.getLength();
  }

  /**
   * <p>
   * the starting position on the chromosome in unit bp (base pairs)
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: false</li>
   * <li>Property is defined in the data-access object ExperimentalCdna</li>
   * </ul>
   * </p>
   * @return a java.lang.Long startPosition
   */
  public java.lang.Long getStartPosition() {
    return dto.getStartPosition();
  }

  /**
   * <p>
   * the ending position on the chromosome in unit bp (base pairs)
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: false</li>
   * <li>Property is defined in the data-access object ExperimentalCdna</li>
   * </ul>
   * </p>
   * @return a java.lang.Long endPosition
   */
  public java.lang.Long getEndPosition() {
    return dto.getEndPosition();
  }

  /**
   * <p>
   * the date on which the protein record was created in the source database
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: false</li>
   * <li>Property is defined in the data-access object ExperimentalCdna</li>
   * </ul>
   * </p>
   * @return a java.sql.Timestamp dateEntered
   */
  public java.sql.Timestamp getDateEntered() {
    return dto.getDateEntered();
  }

  /**
   * <p>
   * the transcript for which the cDNA was experimentally determined
   * </p>
   * <p>
   * Added by AddToOneAssociationRequiredObjectProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: false</li>
   * <li>Property is defined in the data-access object ExperimentalCdna</li>
   * </ul>
   * </p>
   * @return a org.tair.bs.genomicregion.BsTranscript transcript
   */
  public org.tair.bs.genomicregion.BsTranscript getTranscript() {
    // Return 4
    return new org.tair.bs.genomicregion.BsTranscript(dto.getTranscript());
  }
}