/**
 * Copyright 2012 Carnegie Institution for Science. All rights reserved.
 */
// Generated by AndroMDA Poesys/DB cartridge: DO NOT EDIT DIRECTLY.
// Template: AbstractBsDto.vsl


package org.tair.bs.genomicregion;


import com.poesys.bs.delegate.DelegateException;
import com.poesys.bs.dto.IDto;
import com.poesys.db.dto.DtoStatusException;
import com.poesys.db.pk.IPrimaryKey;
import com.poesys.bs.dto.AbstractDto;


import org.tair.db.genomicregion.ITranscript;
import org.tair.db.genomicregion.TranscriptProxy;
import org.tair.db.genomicregion.Transcript;


/**
 * <p>
 * A business layer data-transfer object (DTO) for the Transcript. This class
 * is an abstract class that contains AndroMDA generated code; change nothing
 * in this class. Instead, override any methods in the concrete subclass
 * generated in the same package. AndroMDA will overwrite this class each time
 * you run it but will never overwrite the concrete subclass.
 * </p>
 * <p>
 * SO:0000673: An RNA synthesized on a DNA or RNA template by an RNA polymerase.
 * </p>
 * <p>
 * Stereotypes:
 * </p>
 * <ul>
 *     <li>Persistent</li>
 * </ul>
 * <p>
 * Class is read/write: true
 * </p>
 *
 * @author Poesys/DB Cartridge
 */
public abstract class AbstractBsTranscript 
    extends AbstractDto<org.tair.db.genomicregion.ITranscript> {

  /**
   * A Collection builder helper class for building a business-layer Collection 
   * of BsEst objects from a Collection ests of data-access-layer 
   * org.tair.db.genomicregion.IEst objects
   */
  private class BsEstsCollectionBuilder 
      extends com.poesys.bs.dto.CollectionBuilder<org.tair.db.genomicregion.IEst, org.tair.bs.genomicregion.BsEst> {
    @Override
    public org.tair.bs.genomicregion.BsEst get(org.tair.db.genomicregion.IEst dto) {
      return new org.tair.bs.genomicregion.BsEst(dto);
    }
  }

  /**
   * A Collection builder helper class for building a data-access-layer Collection 
   * of Est objects from an input Collection ests of 
   * business-layer BsEst objects
   */
  private class EstsCollectionBuilder 
      extends com.poesys.bs.dto.CollectionBuilder<org.tair.bs.genomicregion.BsEst, org.tair.db.genomicregion.IEst> {
    @Override
    public org.tair.db.genomicregion.IEst get(org.tair.bs.genomicregion.BsEst dto) {
      return dto.toDto();
    }
  }

  /**
   * A Collection builder helper class for building a business-layer Collection 
   * of BsPolypeptide objects from a Collection proteins of data-access-layer 
   * org.tair.db.genomicregion.IPolypeptide objects
   */
  private class BsProteinsCollectionBuilder 
      extends com.poesys.bs.dto.CollectionBuilder<org.tair.db.genomicregion.IPolypeptide, org.tair.bs.genomicregion.BsPolypeptide> {
    @Override
    public org.tair.bs.genomicregion.BsPolypeptide get(org.tair.db.genomicregion.IPolypeptide dto) {
      return new org.tair.bs.genomicregion.BsPolypeptide(dto);
    }
  }

  /**
   * A Collection builder helper class for building a data-access-layer Collection 
   * of Polypeptide objects from an input Collection proteins of 
   * business-layer BsPolypeptide objects
   */
  private class ProteinsCollectionBuilder 
      extends com.poesys.bs.dto.CollectionBuilder<org.tair.bs.genomicregion.BsPolypeptide, org.tair.db.genomicregion.IPolypeptide> {
    @Override
    public org.tair.db.genomicregion.IPolypeptide get(org.tair.bs.genomicregion.BsPolypeptide dto) {
      return dto.toDto();
    }
  }

  /**
   * A Collection builder helper class for building a business-layer Collection 
   * of BsExperimentalCdna objects from a Collection experimentalCdnas of data-access-layer 
   * org.tair.db.genomicregion.IExperimentalCdna objects
   */
  private class BsExperimentalCdnasCollectionBuilder 
      extends com.poesys.bs.dto.CollectionBuilder<org.tair.db.genomicregion.IExperimentalCdna, org.tair.bs.genomicregion.BsExperimentalCdna> {
    @Override
    public org.tair.bs.genomicregion.BsExperimentalCdna get(org.tair.db.genomicregion.IExperimentalCdna dto) {
      return new org.tair.bs.genomicregion.BsExperimentalCdna(dto);
    }
  }

  /**
   * A Collection builder helper class for building a data-access-layer Collection 
   * of ExperimentalCdna objects from an input Collection experimentalCdnas of 
   * business-layer BsExperimentalCdna objects
   */
  private class ExperimentalCdnasCollectionBuilder 
      extends com.poesys.bs.dto.CollectionBuilder<org.tair.bs.genomicregion.BsExperimentalCdna, org.tair.db.genomicregion.IExperimentalCdna> {
    @Override
    public org.tair.db.genomicregion.IExperimentalCdna get(org.tair.bs.genomicregion.BsExperimentalCdna dto) {
      return dto.toDto();
    }
  }

  /**
   * Create a BsTranscript object from a Transcript object.
   * 
   * @param dto the data-access layer Transcript DTO
   * @throws DelegateException when there is a problem creating the Transcript
   */
  public AbstractBsTranscript(ITranscript dto) throws DelegateException {
    super(dto);
  }

  /**
   * Create a Transcript from new data.
   *
   * @param key the primary key of the Transcript
   * @param tairObjectId the unique identifier in the TAIR database for the region (locus, gene,
chromosome, contig)
   * @param name the name of the genomic region, such as the chromosome name (1, M), the
locus/gene name (AT1G01010), the transcript name (AT1G01010.1), or the contig
name (MI421-MI421); may or may not be unique within a genome
   * @param dateEntered the date on which the region was created in the source database
   * @param parentRegionTairObjectId foreign key used by setter to query associated object
   * @param genome foreign key used by setter to query associated object
   * @param sequence the complete DNA sequence for the transcript
   * @param seqTairObjectId the unique identifier for the cDNA
   * @param genbankAccession the unique identifier at GenBank for the cDNA sequence
   * @param primarySymbol the primary symbolic name for the transcript
   * @param primarySymbolFullName the full name for the primary symbolic name for the transcript (may be null)
   * @param startPosition the start of the transcript on the top level region (chromosome or contig)
   * @param endPosition the end of the transcript on the top-level parent region (chromosome or contig)
   * @param length the length of the transcript genomic region in base pairs (bp)
   * @param orientation the direction of the transcript genomic region in the DNA: forward or reverse;
   * @param basis the basis for assigning the transcript to the map:
BAC_end_sequencing
recombination
gene_annotation
fingerprint
hybridization
unknown
sequence_match
fingerprint_and_hybridization
   * @param derived whether the transcript  is derived ab initio or experimentally known in vivo
   * @param representativeGeneModel Is this transcript the representative alternatively spliced transcript for the
gene?
   * @param transcribesTairObjectId foreign key used by setter to query associated object
   */
  public AbstractBsTranscript(IPrimaryKey key, java.lang.Long tairObjectId, java.lang.String name, java.sql.Timestamp dateEntered, java.lang.Long parentRegionTairObjectId, java.lang.String genome, java.lang.String sequence, java.lang.Long seqTairObjectId, java.lang.String genbankAccession, java.lang.String primarySymbol, java.lang.String primarySymbolFullName, java.lang.Long startPosition, java.lang.Long endPosition, java.lang.Long length, java.lang.String orientation, java.lang.String basis, java.lang.Boolean derived, java.lang.Boolean representativeGeneModel, java.lang.Long transcribesTairObjectId) {
    super(new TranscriptProxy(new Transcript(key, tairObjectId, name, dateEntered, parentRegionTairObjectId, genome, sequence, seqTairObjectId, genbankAccession, primarySymbol, primarySymbolFullName, startPosition, endPosition, length, orientation, basis, derived, representativeGeneModel, transcribesTairObjectId)));
  }

  @SuppressWarnings("unchecked")
  @Override
  public boolean equals(Object arg0) {
    // Unchecked cast here
    IDto<ITranscript> other = (IDto<ITranscript>)arg0;
    return dto.equals(other.toDto());
  }

  @Override
  public int hashCode() {
    return dto.hashCode();
  }

  public int compareTo(IDto<ITranscript> o) {
    return dto.compareTo(o.toDto());
  }

  public void markChildrenDeleted() throws DtoStatusException {
    dto.markChildrenDeleted();
  }
  
  public IPrimaryKey getPrimaryKey() {
    return dto.getPrimaryKey();
  }

  // Inherited properties

  /**
   * <p>
   * the unique identifier in the TAIR database for the region (locus, gene,
   * chromosome, contig)
   * </p>
   * <p>
   * Added by AddNaturalKeyProperties + AddInheritedKeyProperties through inheritance
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: false</li>
   * <li>Property is inherited from a superclass</li>
   * </ul>
   * </p>
   * @return a java.lang.Long tairObjectId
   */

  public java.lang.Long getTairObjectId() {
    return dto.getTairObjectId();
  }
// Setter here if the four conditions below are all true or not false
// Read/Write DTO: true
// Mutable DTO: not false
// Mutable property: not false
// Read/Write property: false

  /**
   * 
   * <p>
   * Added by addAssociationObjects + AddInheritedAssociationObjectProperties through inheritance
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is inherited from a superclass</li>
   * </ul>
   * </p>
   * @return a org.tair.db.genomicregion.IRegion parent
   */
  public org.tair.bs.genomicregion.BsRegion getParent() {
    org.tair.bs.genomicregion.BsRegion bsObject = null;
    org.tair.db.genomicregion.IRegion object = dto.getParent();
	if (object != null) {
	  bsObject = new org.tair.bs.genomicregion.BsRegion(object);
	}
	return bsObject;
  }
// Setter here if the four conditions below are all true or not false
// Read/Write DTO: true
// Mutable DTO: not false
// Mutable property: not false
// Read/Write property: true

  /**
   * <p>
   * Set the parent.
   * </p>
   * 
   * <p>
   * Added by addAssociationObjects + AddInheritedAssociationObjectProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is inherited from a superclass</li>
   * </ul>
   * </p>
   * @param parent the value to set into the object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   */
  public void setParent(org.tair.bs.genomicregion.BsRegion parent) 
      throws com.poesys.db.dto.DtoStatusException{
    dto.setParent(parent.toDto());
  }

  /**
   * 
   * <p>
   * Added by addAssociationObjects + AddInheritedAssociationObjectProperties through inheritance
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is inherited from a superclass</li>
   * </ul>
   * </p>
   * @return a org.tair.db.genomicregion.IReferenceGenome referenceGenome
   */
  public org.tair.bs.genomicregion.BsReferenceGenome getReferenceGenome() {
    // Return 2
    return new org.tair.bs.genomicregion.BsReferenceGenome(dto.getReferenceGenome());
  }
// Setter here if the four conditions below are all true or not false
// Read/Write DTO: true
// Mutable DTO: not false
// Mutable property: not false
// Read/Write property: true

  /**
   * <p>
   * Set the referenceGenome.
   * </p>
   * 
   * <p>
   * Added by addAssociationObjects + AddInheritedAssociationObjectProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is inherited from a superclass</li>
   * </ul>
   * </p>
   * @param referenceGenome the value to set into the object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   * @throws com.poesys.db.InvalidParametersException when the property is
   *                 required but the input parameter referenceGenome is null
   */
  public void setReferenceGenome(org.tair.bs.genomicregion.BsReferenceGenome referenceGenome) 
      throws com.poesys.db.dto.DtoStatusException, com.poesys.db.InvalidParametersException {
    dto.setReferenceGenome(referenceGenome.toDto());
  }

  /**
   * <p>
   * the name of the genomic region, such as the chromosome name (1, M), the
   * locus/gene name (AT1G01010), the transcript name (AT1G01010.1), or the contig
   * name (MI421-MI421); may or may not be unique within a genome
   * </p>
   * <p>
   * Added by TransformToProperty + AddInheritedAttributeProperties through inheritance
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is inherited from a superclass</li>
   * </ul>
   * </p>
   * @return a java.lang.String name
   */

  public java.lang.String getName() {
    return dto.getName();
  }
// Setter here if the four conditions below are all true or not false
// Read/Write DTO: true
// Mutable DTO: not false
// Mutable property: not false
// Read/Write property: true

  /**
   * <p>
   * Set the name.
   * </p>
   * <p>
   * the name of the genomic region, such as the chromosome name (1, M), the
   * locus/gene name (AT1G01010), the transcript name (AT1G01010.1), or the contig
   * name (MI421-MI421); may or may not be unique within a genome
   * </p>
   * <p>
   * Added by TransformToProperty + AddInheritedAttributeProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is inherited from a superclass</li>
   * </ul>
   * </p>
   * @param name the value to set into the object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   * @throws com.poesys.db.InvalidParametersException when the property is
   *                 required but the input parameter name is null
   */
  public void setName(java.lang.String name) 
      throws com.poesys.db.dto.DtoStatusException, com.poesys.db.InvalidParametersException {
    dto.setName(name);
  }

  /**
   * <p>
   * the date on which the region was created in the source database
   * </p>
   * <p>
   * Added by TransformToProperty + AddInheritedAttributeProperties through inheritance
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is inherited from a superclass</li>
   * </ul>
   * </p>
   * @return a java.sql.Timestamp dateEntered
   */

  public java.sql.Timestamp getDateEntered() {
    return dto.getDateEntered();
  }
// Setter here if the four conditions below are all true or not false
// Read/Write DTO: true
// Mutable DTO: not false
// Mutable property: not false
// Read/Write property: true

  /**
   * <p>
   * Set the dateEntered.
   * </p>
   * <p>
   * the date on which the region was created in the source database
   * </p>
   * <p>
   * Added by TransformToProperty + AddInheritedAttributeProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is inherited from a superclass</li>
   * </ul>
   * </p>
   * @param dateEntered the value to set into the object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   * @throws com.poesys.db.InvalidParametersException when the property is
   *                 required but the input parameter dateEntered is null
   */
  public void setDateEntered(java.sql.Timestamp dateEntered) 
      throws com.poesys.db.dto.DtoStatusException, com.poesys.db.InvalidParametersException {
    dto.setDateEntered(dateEntered);
  }

  /**
   * <p>
   * Foreign key used by setter to query associated object
   * </p>
   * <p>
   * Added by AddNaturalKeyProperties + AddInheritedAssociationAttributeProperties through inheritance
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: false</li>
   * <li>Property is inherited from a superclass</li>
   * </ul>
   * </p>
   * @return a java.lang.Long parentRegionTairObjectId
   */

  public java.lang.Long getParentRegionTairObjectId() {
    return dto.getParentRegionTairObjectId();
  }
// Setter here if the four conditions below are all true or not false
// Read/Write DTO: true
// Mutable DTO: not false
// Mutable property: not false
// Read/Write property: false

  /**
   * <p>
   * Foreign key used by setter to query associated object
   * </p>
   * <p>
   * Added by AddNaturalKeyProperties + AddInheritedAssociationAttributeProperties through inheritance
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: false</li>
   * <li>Property is inherited from a superclass</li>
   * </ul>
   * </p>
   * @return a java.lang.String genome
   */

  public java.lang.String getGenome() {
    return dto.getGenome();
  }
// Setter here if the four conditions below are all true or not false
// Read/Write DTO: true
// Mutable DTO: not false
// Mutable property: not false
// Read/Write property: false

  // Data member properties

  /**
   * <p>
   * the complete DNA sequence for the transcript
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @return a java.lang.String sequence
   */
  public java.lang.String getSequence() {
    return dto.getSequence();
  }

  /**
   * <p>
   * Set the sequence.
   * </p>
   * <p>
   * the complete DNA sequence for the transcript
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @param sequence the associated business object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   * @throws com.poesys.db.InvalidParametersException when the property is
   *                 required but the input parameter sequence is null
   */
  public void setSequence(java.lang.String sequence) 
      throws com.poesys.db.dto.DtoStatusException , com.poesys.db.InvalidParametersException {
    dto.setSequence(sequence);
  }

  /**
   * <p>
   * the unique identifier for the cDNA
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @return a java.lang.Long seqTairObjectId
   */
  public java.lang.Long getSeqTairObjectId() {
    return dto.getSeqTairObjectId();
  }

  /**
   * <p>
   * Set the seqTairObjectId.
   * </p>
   * <p>
   * the unique identifier for the cDNA
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @param seqTairObjectId the associated business object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   * @throws com.poesys.db.InvalidParametersException when the property is
   *                 required but the input parameter seqTairObjectId is null
   */
  public void setSeqTairObjectId(java.lang.Long seqTairObjectId) 
      throws com.poesys.db.dto.DtoStatusException , com.poesys.db.InvalidParametersException {
    dto.setSeqTairObjectId(seqTairObjectId);
  }

  /**
   * <p>
   * the unique identifier at GenBank for the cDNA sequence
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @return a java.lang.String genbankAccession
   */
  public java.lang.String getGenbankAccession() {
    return dto.getGenbankAccession();
  }

  /**
   * <p>
   * Set the genbankAccession.
   * </p>
   * <p>
   * the unique identifier at GenBank for the cDNA sequence
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @param genbankAccession the associated business object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   */
  public void setGenbankAccession(java.lang.String genbankAccession) 
      throws com.poesys.db.dto.DtoStatusException {
    dto.setGenbankAccession(genbankAccession);
  }

  /**
   * <p>
   * the primary symbolic name for the transcript
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @return a java.lang.String primarySymbol
   */
  public java.lang.String getPrimarySymbol() {
    return dto.getPrimarySymbol();
  }

  /**
   * <p>
   * Set the primarySymbol.
   * </p>
   * <p>
   * the primary symbolic name for the transcript
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @param primarySymbol the associated business object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   */
  public void setPrimarySymbol(java.lang.String primarySymbol) 
      throws com.poesys.db.dto.DtoStatusException {
    dto.setPrimarySymbol(primarySymbol);
  }

  /**
   * <p>
   * the full name for the primary symbolic name for the transcript (may be null)
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @return a java.lang.String primarySymbolFullName
   */
  public java.lang.String getPrimarySymbolFullName() {
    return dto.getPrimarySymbolFullName();
  }

  /**
   * <p>
   * Set the primarySymbolFullName.
   * </p>
   * <p>
   * the full name for the primary symbolic name for the transcript (may be null)
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @param primarySymbolFullName the associated business object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   */
  public void setPrimarySymbolFullName(java.lang.String primarySymbolFullName) 
      throws com.poesys.db.dto.DtoStatusException {
    dto.setPrimarySymbolFullName(primarySymbolFullName);
  }

  /**
   * <p>
   * the start of the transcript on the top level region (chromosome or contig)
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @return a java.lang.Long startPosition
   */
  public java.lang.Long getStartPosition() {
    return dto.getStartPosition();
  }

  /**
   * <p>
   * Set the startPosition.
   * </p>
   * <p>
   * the start of the transcript on the top level region (chromosome or contig)
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @param startPosition the associated business object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   * @throws com.poesys.db.InvalidParametersException when the property is
   *                 required but the input parameter startPosition is null
   */
  public void setStartPosition(java.lang.Long startPosition) 
      throws com.poesys.db.dto.DtoStatusException , com.poesys.db.InvalidParametersException {
    dto.setStartPosition(startPosition);
  }

  /**
   * <p>
   * the end of the transcript on the top-level parent region (chromosome or contig)
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @return a java.lang.Long endPosition
   */
  public java.lang.Long getEndPosition() {
    return dto.getEndPosition();
  }

  /**
   * <p>
   * Set the endPosition.
   * </p>
   * <p>
   * the end of the transcript on the top-level parent region (chromosome or contig)
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @param endPosition the associated business object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   * @throws com.poesys.db.InvalidParametersException when the property is
   *                 required but the input parameter endPosition is null
   */
  public void setEndPosition(java.lang.Long endPosition) 
      throws com.poesys.db.dto.DtoStatusException , com.poesys.db.InvalidParametersException {
    dto.setEndPosition(endPosition);
  }

  /**
   * <p>
   * the length of the transcript genomic region in base pairs (bp)
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @return a java.lang.Long length
   */
  public java.lang.Long getLength() {
    return dto.getLength();
  }

  /**
   * <p>
   * Set the length.
   * </p>
   * <p>
   * the length of the transcript genomic region in base pairs (bp)
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @param length the associated business object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   * @throws com.poesys.db.InvalidParametersException when the property is
   *                 required but the input parameter length is null
   */
  public void setLength(java.lang.Long length) 
      throws com.poesys.db.dto.DtoStatusException , com.poesys.db.InvalidParametersException {
    dto.setLength(length);
  }

  /**
   * <p>
   * the direction of the transcript genomic region in the DNA: forward or reverse;
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @return a java.lang.String orientation
   */
  public java.lang.String getOrientation() {
    return dto.getOrientation();
  }

  /**
   * <p>
   * Set the orientation.
   * </p>
   * <p>
   * the direction of the transcript genomic region in the DNA: forward or reverse;
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @param orientation the associated business object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   * @throws com.poesys.db.InvalidParametersException when the property is
   *                 required but the input parameter orientation is null
   */
  public void setOrientation(java.lang.String orientation) 
      throws com.poesys.db.dto.DtoStatusException , com.poesys.db.InvalidParametersException {
    dto.setOrientation(orientation);
  }

  /**
   * <p>
   * the basis for assigning the transcript to the map:
   * </p>
   * <p>
   * BAC_end_sequencing
   * </p>
   * <p>
   * recombination
   * </p>
   * <p>
   * gene_annotation
   * </p>
   * <p>
   * fingerprint
   * </p>
   * <p>
   * hybridization
   * </p>
   * <p>
   * unknown
   * </p>
   * <p>
   * sequence_match
   * </p>
   * <p>
   * fingerprint_and_hybridization
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @return a java.lang.String basis
   */
  public java.lang.String getBasis() {
    return dto.getBasis();
  }

  /**
   * <p>
   * Set the basis.
   * </p>
   * <p>
   * the basis for assigning the transcript to the map:
   * </p>
   * <p>
   * BAC_end_sequencing
   * </p>
   * <p>
   * recombination
   * </p>
   * <p>
   * gene_annotation
   * </p>
   * <p>
   * fingerprint
   * </p>
   * <p>
   * hybridization
   * </p>
   * <p>
   * unknown
   * </p>
   * <p>
   * sequence_match
   * </p>
   * <p>
   * fingerprint_and_hybridization
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @param basis the associated business object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   * @throws com.poesys.db.InvalidParametersException when the property is
   *                 required but the input parameter basis is null
   */
  public void setBasis(java.lang.String basis) 
      throws com.poesys.db.dto.DtoStatusException , com.poesys.db.InvalidParametersException {
    dto.setBasis(basis);
  }

  /**
   * <p>
   * whether the transcript  is derived ab initio or experimentally known in vivo
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @return a java.lang.Boolean derived
   */
  public java.lang.Boolean getDerived() {
    return dto.getDerived();
  }

  /**
   * <p>
   * Set the derived.
   * </p>
   * <p>
   * whether the transcript  is derived ab initio or experimentally known in vivo
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @param derived the associated business object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   * @throws com.poesys.db.InvalidParametersException when the property is
   *                 required but the input parameter derived is null
   */
  public void setDerived(java.lang.Boolean derived) 
      throws com.poesys.db.dto.DtoStatusException , com.poesys.db.InvalidParametersException {
    dto.setDerived(derived);
  }

  /**
   * <p>
   * Is this transcript the representative alternatively spliced transcript for the
   * gene?
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @return a java.lang.Boolean representativeGeneModel
   */
  public java.lang.Boolean getRepresentativeGeneModel() {
    return dto.getRepresentativeGeneModel();
  }

  /**
   * <p>
   * Set the representativeGeneModel.
   * </p>
   * <p>
   * Is this transcript the representative alternatively spliced transcript for the
   * gene?
   * </p>
   * <p>
   * Added by AddLocalAttributeProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @param representativeGeneModel the associated business object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   * @throws com.poesys.db.InvalidParametersException when the property is
   *                 required but the input parameter representativeGeneModel is null
   */
  public void setRepresentativeGeneModel(java.lang.Boolean representativeGeneModel) 
      throws com.poesys.db.dto.DtoStatusException , com.poesys.db.InvalidParametersException {
    dto.setRepresentativeGeneModel(representativeGeneModel);
  }

  /**
   * 
   * <p>
   * Added by AddToOneAssociationRequiredObjectProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @return a org.tair.bs.genomicregion.BsGene transcribes
   */
  public org.tair.bs.genomicregion.BsGene getTranscribes() {
    // Return 4
    return new org.tair.bs.genomicregion.BsGene(dto.getTranscribes());
  }

  /**
   * <p>
   * Set the transcribes.
   * </p>
   * 
   * <p>
   * Added by AddToOneAssociationRequiredObjectProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @param transcribes the associated business object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   * @throws com.poesys.db.InvalidParametersException when the property is
   *                 required but the input parameter transcribes is null
   */
  public void setTranscribes(org.tair.bs.genomicregion.BsGene transcribes) 
      throws com.poesys.db.dto.DtoStatusException , com.poesys.db.InvalidParametersException {
    dto.setTranscribes(transcribes.toDto());
  }

  /**
   * <p>
   * Foreign key used by setter to query associated object
   * </p>
   * <p>
   * Added by AddNaturalKeyProperties + AddInheritedKeyProperties + AddToOneAssociationAttributeProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: false</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @return a java.lang.Long transcribesTairObjectId
   */
  public java.lang.Long getTranscribesTairObjectId() {
    return dto.getTranscribesTairObjectId();
  }

  /**
   * <p>
   * the collection of Extended Sequence Tags associated with the transcript
   * </p>
   * <p>
   * Added by AddToManyChildCollectionProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @return a org.tair.bs.genomicregion.BsEst ests
   */
  public java.util.Collection<org.tair.bs.genomicregion.BsEst> getEsts() {
    BsEstsCollectionBuilder builder = new BsEstsCollectionBuilder();
    return builder.getCollection(dto.getEsts());
  }

  /**
   * <p>
   * Set the ests.
   * </p>
   * <p>
   * the collection of Extended Sequence Tags associated with the transcript
   * </p>
   * <p>
   * Added by AddToManyChildCollectionProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @param ests the associated business object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   */
  public void setEsts(java.util.Collection<org.tair.bs.genomicregion.BsEst> ests) 
      throws com.poesys.db.dto.DtoStatusException{
    EstsCollectionBuilder builder = new EstsCollectionBuilder();
      dto.setEsts(builder.getCollection(ests));
  }

  /**
   * <p>
   * the collection of polypeptides or proteins coded for by the transcript
   * </p>
   * <p>
   * Added by AddToManyChildCollectionProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @return a org.tair.bs.genomicregion.BsPolypeptide proteins
   */
  public java.util.Collection<org.tair.bs.genomicregion.BsPolypeptide> getProteins() {
    BsProteinsCollectionBuilder builder = new BsProteinsCollectionBuilder();
    return builder.getCollection(dto.getProteins());
  }

  /**
   * <p>
   * Set the proteins.
   * </p>
   * <p>
   * the collection of polypeptides or proteins coded for by the transcript
   * </p>
   * <p>
   * Added by AddToManyChildCollectionProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @param proteins the associated business object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   */
  public void setProteins(java.util.Collection<org.tair.bs.genomicregion.BsPolypeptide> proteins) 
      throws com.poesys.db.dto.DtoStatusException{
    ProteinsCollectionBuilder builder = new ProteinsCollectionBuilder();
      dto.setProteins(builder.getCollection(proteins));
  }

  /**
   * <p>
   * the collection of cDNAs experimentally derived in vivo for the transcript
   * </p>
   * <p>
   * Added by AddToManyChildCollectionProperties as data member
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @return a org.tair.bs.genomicregion.BsExperimentalCdna experimentalCdnas
   */
  public java.util.Collection<org.tair.bs.genomicregion.BsExperimentalCdna> getExperimentalCdnas() {
    BsExperimentalCdnasCollectionBuilder builder = new BsExperimentalCdnasCollectionBuilder();
    return builder.getCollection(dto.getExperimentalCdnas());
  }

  /**
   * <p>
   * Set the experimentalCdnas.
   * </p>
   * <p>
   * the collection of cDNAs experimentally derived in vivo for the transcript
   * </p>
   * <p>
   * Added by AddToManyChildCollectionProperties
   * </p>
   * <p>
   * <ul>
   * <li>Property is read/write: true</li>
   * <li>Property is defined in the data-access object Transcript</li>
   * </ul>
   * </p>
   * @param experimentalCdnas the associated business object
   * @throws com.poesys.db.dto.DtoStatusException when the system can't set
   *                 the data-access status to CHANGED
   */
  public void setExperimentalCdnas(java.util.Collection<org.tair.bs.genomicregion.BsExperimentalCdna> experimentalCdnas) 
      throws com.poesys.db.dto.DtoStatusException{
    ExperimentalCdnasCollectionBuilder builder = new ExperimentalCdnasCollectionBuilder();
      dto.setExperimentalCdnas(builder.getCollection(experimentalCdnas));
  }

  /**
   * Add a Est object to the ests collection.
   *
   * @param object the object to add to the collection
   */
  public void addEstsEst(org.tair.bs.genomicregion.BsEst object) {
    dto.addEstsEst(object.toDto());
  }

  /**
   * Add a Polypeptide object to the proteins collection.
   *
   * @param object the object to add to the collection
   */
  public void addProteinsPolypeptide(org.tair.bs.genomicregion.BsPolypeptide object) {
    dto.addProteinsPolypeptide(object.toDto());
  }

  /**
   * Add a ExperimentalCdna object to the experimentalCdnas collection.
   *
   * @param object the object to add to the collection
   */
  public void addExperimentalCdnasExperimentalCdna(org.tair.bs.genomicregion.BsExperimentalCdna object) {
    dto.addExperimentalCdnasExperimentalCdna(object.toDto());
  }
}